### DISCUSSION TITLE
Release 5.0 and beyond

### DISCUSSION THREAD

**Discussion by mariusconjeaud** [neo4j-contrib/neomodel#689] [Announcements]
Dear neomodel community,

As you might have seen in the past weeks, activity has resumed for neomodel.

### Latest updates
The first step has been to update the existing version (4.x) to support all of Neo4j 4.x versions (up to the LTS 4.4).

The second step has been migrating the packaging and deployment to more recent solutions (pyproject.toml, GitHub Actions,...) along with some code cleanup (prettifying, linting, ...)

Finally, the latest PRs merged upped the support to Neo4j 5.x and Python 3.11
**Please note that for now, the master branch is the very latest, meaning it supports Neo4j 5.x and 4.4 ; whereas the latest release in PyPI is still 4.0 which stops at Neo4j version 4.4** To avoid this confusion in the future, we will be creating PRs to "release candidate" branches, so that the master branch always matches the latest PyPI release.

### Future updates
Since we are moving again with neomodel, which had been standing quite still for a couple years, we would like your opinion in what you would like to see coming in the next releases ; and how you would like to contribute.

For this, please post in the "Ideas" section of the Discussions. You can also pick topics from the [TODO Wiki](https://github.com/neo4j-contrib/neomodel/wiki/TODOs-&-Enhancements) that you would especially like. _Note, this hasn't been updated since 2019, but most of the ideas there have probably not been implemented since_

Thanks a lot and I hope to hear from you in the coming months !

**Comment by mariusconjeaud** [comment]
Neomodel [version 5.0.0 is out now](https://pypi.org/project/neomodel/) !

### Things to note
This release does not introduce breaking changes, except in the versions of Neo4j (and Python in certain cases) that it supports :

- All, and only, officially supported versions of Neo4j are now supported by neomodel : this means 4.4 (LTS) and 5.x are supported, but _not 4.3 and before_
- This is also true for Python : versions 3.7 through 3.11 are supported. Python 3.6 support was dropped in neomodel version 4.0.10 released in February 2023
- Relationship indexes were added

### Full changelog

- Confirmed support of Neo4j versions 5.x and 4.4 (LTS)
- Dropped support of EOL Neo4j versions (4.3 and below)
- Confirmed support of Python 3.11
- Migrated RelationshipTo/RelationshipFrom from a method to a class for consistency.
- Add support for relationship indexes
- Auto-generated indexes now have a deterministic name
- Code cleanup

### Next steps
Now that neomodel supports a version of Neo4j that will be supported for some years still, we have some breathing space to implement bigger features that can bring a lot of value to neomodel. We already have some candidates for this, but we also intend to go through the Issues panel, and clean it up as much as possible by finding out which ones can bring value to neomodel. As per my previous message, please post in the "Ideas" section of the Discussions. You can also pick topics from the [TODO Wiki](https://github.com/neo4j-contrib/neomodel/wiki/TODOs-&-Enhancements) that you would especially like.

Enjoy, and hope to hear from you !

**Reply by jpbianchi** [reply]
I am pushing for Neomodel at work but my colleagues are worried because the documentation is not very detailed and we can't find detailed examples we could learn from.  Is there a community where we can ask questions / learn from?

**Reply by mariusconjeaud** [reply]
There isn't a community space beyond this GitHub space, maybe that is worth considering.

As for examples, the way I learned the most when picking up maintenance of neomodel was through two means :

- [Reading the tests](https://github.com/neo4j-contrib/neomodel/tree/master/test/sync_)
- The [OpenStudyBuilder](https://gitlab.com/Novo-Nordisk/nn-public/openstudybuilder/OpenStudyBuilder-Solution/-/tree/main/clinical-mdr-api/clinical_mdr_api) project, which I am involved in (and how I ended up maintaining neomodel). You can find the Model definitions in [domain_repositories/models](https://gitlab.com/Novo-Nordisk/nn-public/openstudybuilder/OpenStudyBuilder-Solution/-/tree/main/clinical-mdr-api/clinical_mdr_api/domain_repositories/models) and the logic using neomodel in [domain_repositories](https://gitlab.com/Novo-Nordisk/nn-public/openstudybuilder/OpenStudyBuilder-Solution/-/tree/main/clinical-mdr-api/clinical_mdr_api/domain_repositories/models) itself. This project makes use of most of neomodel's features, except the bulk import part, and it can be interesting to show how to set up a project, use the models, advanced querying, etc...

**Reply by jpbianchi** [reply]
Thanks @mariusconjeaud.
At the moment, I'm just trying to start from a node, traverse several relationships and collect properties along the way, but I just can't find the right syntax.  
It would help if the doc could give examples with the full capabilities of every method: traverse_, annotate, etc
 I'll take a look at your project for now, but a Discord group would be a great help I'm sure.

**Reply by mariusconjeaud** [reply]
Thanks for the feedback. Always hard to maintain documentation, especially as here it has a long history with most of it not being written by current maintainers. So we just "go along" with was already there, and it's hard to think of new ways to write it. Anyway !

For traverse_relations, did you see [this section ](https://neomodel.readthedocs.io/en/latest/traversal.html) ?
Also, the Getting Started guide shows [a full example with all the advanced methods](https://neomodel.readthedocs.io/en/latest/getting_started.html#full-example) tied together, and the corresponding Cypher query, to show how they interact. Not saying this is enough, but I thought it might help (a bit).
Or is your problem more on how to parse the results of those ?

I will look at the Discord option

**Comment by jpbianchi** [comment]
Thanks Marius,

My issue is to traverse several relationships and be able to collect
properties from the nodes traversed.

For instance, if I start with the Company node, then the CreditScore, then
the data, then the date, with
.traverse_relations(d=company__creditscore__data__date), I don't know how
to get the company name, creditscore description, data value since the
alias d points to the date node, so I can retrieve only the year and month.
I'll take a look at the provided examples.

Thank you very much
